### PR TITLE
[WIP] Implement empty-tensor operators for XPU backend

### PR DESCRIPTION
**Pull request by fengyuan14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119157
* #117234
* __->__ #118073
* #117970

Features: https://github.com/pytorch/pytorch/issues/114835

1. Enhance building capability for `libtorch_xpu.so` with `sycl_add_library`. `libtorch_xpu.so` is XPU backend implementation of PyTorch ATen tensor library, including
   kernel part in SYCL and host part in C++. Enable separate compilation by distinguishing SYCL and C++ sources and using SYCL and C++ compilers compile them separately.
2. Add two Tensor factory ATen operators for XPU backend.

Signed-off-by: Feng Yuan <feng1.yuan@intel.com>

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by arthuryuan1987** [comment]
Convert to Draft since the PR depends on [Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116869)

File: test/test_xpu.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+from torch.testing._internal.common_utils import TestCase, \
+    run_tests, IS_WINDOWS
+from torch.testing._internal.common_device_type import (
+    instantiate_device_type_tests,
+)
+
+# not torch.backends.xpu.is_avaliable
+@unittest.skipIf(IS_WINDOWS, "XPU operators testing is disabled")
+class TestXPUOperators(TestCase):
+    def test_tensor_factory(self):
```
Comment (jgong5) [inline test/test_xpu.py:15]: Suggest not to pack multiple tests into a single UT. Separate them and name them explicity with `test_empty` and `test_empty_strided`.

File: aten/src/ATen/xpu/EmptyTensor.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
+#pragma once
+#include <ATen/core/TensorBase.h>
+
+namespace at::detail {
+
+// XXX: add TORCH_XPU_API
```
Comment (jgong5) [inline aten/src/ATen/xpu/EmptyTensor.h:6]: Do we need to expose them as API?

File: aten/src/ATen/native/xpu/TensorFactories.cpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+#include <ATen/ops/empty_native.h>
+#include <ATen/ops/empty_strided_native.h>
+#endif
+
+namespace at::native {
+
+Tensor empty_xpu(IntArrayRef size, c10::optional<ScalarType> dtype_opt, c10::optional<Layout> layout_opt, c10::optional<Device> device_opt, c10::optional<bool> pin_memory_opt, c10::optional<c10::MemoryFormat> memory_format_opt) {
+  Tensor result = at::detail::empty_xpu(size, dtype_opt, layout_opt, device_opt, pin_memory_opt, memory_format_opt);
+  // See Note [Enabling Deterministic Operations]
+  TORCH_CHECK(!(C10_UNLIKELY(at::globalContext().deterministicAlgorithms() && at::globalContext().deterministicFillUninitializedMemory())), "XPU backend doesn't support deterministic implementation for empty ...")
```
Comment (jgong5) [inline aten/src/ATen/native/xpu/TensorFactories.cpp:19]: May I know why `empty` call is related to determinism? It doesn't initialize memory, right?

File: aten/src/ATen/xpu/EmptyTensor.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+TensorBase empty_xpu(
+    IntArrayRef size,
+    ScalarType dtype,
+    c10::optional<Device> device_opt,
+    c10::optional<c10::MemoryFormat> memory_format_opt) {
+  const auto device = device_or_default(device_opt);
+  TORCH_INTERNAL_ASSERT(device.is_xpu());
+  // XXX
+  // const c10::DeviceGuard device_guard(device);
+  auto* allocator = at::getCPUAllocator();
```
Comment (jgong5) [inline aten/src/ATen/xpu/EmptyTensor.cpp:18]: Hack... Are you going to fix it in this PR?

File: aten/src/ATen/xpu/EmptyTensor.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+TensorBase empty_strided_xpu(
+    IntArrayRef size,
+    IntArrayRef stride,
+    ScalarType dtype,
+    c10::optional<Device> device_opt) {
+  const auto device = device_or_default(device_opt);
+  TORCH_INTERNAL_ASSERT(device.is_xpu());
+  // XXX
+  // const c10::DeviceGuard device_guard(device);
+  auto* allocator = at::getCPUAllocator();
```
Comment (jgong5) [inline aten/src/ATen/xpu/EmptyTensor.cpp:60]: ditto

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/xpu/EmptyTensor.cpp (hunk 4 above)
Comment (arthuryuan1987) [reply aten/src/ATen/xpu/EmptyTensor.cpp:18]: Yes. It depends on runtime PRs landed.

File: aten/src/ATen/xpu/EmptyTensor.h (hunk 2 above)
Comment (arthuryuan1987) [reply aten/src/ATen/xpu/EmptyTensor.h:6]: Yes, here we will follow CUDA exposure.

File: aten/src/ATen/native/xpu/TensorFactories.cpp (hunk 3 above)
Comment (jgong5) [reply aten/src/ATen/native/xpu/TensorFactories.cpp:19]: Hmm, I checked cuda implementation `empty_cuda` and didn't find the checking of deterministic algorithms. Do I miss anything?

Comment (arthuryuan1987) [reply aten/src/ATen/native/xpu/TensorFactories.cpp:19]: Yes, I will retrieve it back, if we have fill kernel.

**Pull request closed without merging**